src/Plugin/PasswordConstraint/Number.php

```diff
@@ -0,0 +1,41 @@
+<?php
+
+namespace Drupal\password_enhancements\Plugin\PasswordConstraint;
+
+use Drupal\password_enhancements\Plugin\PasswordConstraintPluginInterface;
+
+/**
+ * Number password constraint plugin.
+ *
+ * @PasswordConstraint(
+ *   id = "number",
+ *   name = @Translation("Number"),
+ *   description = @Translation("Checks if the password has at least a specified number of numbers."),
+ *   unique = TRUE,
+ *   jsLibrary = "password_enhancements/plugin.number",
+ * )
+ */
+final class Number extends MinimumCharacters implements PasswordConstraintPluginInterface {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function defaultDescriptionSingular(): string {
+    return 'Add at least one number.';
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function defaultDescriptionPlural(): string {
+    return 'Add @minimum_characters more numbers.';
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function validate(string $value): bool {
+    return parent::validateValue(preg_replace('/([^0-9])/', '', $value));
```

Comment:
Do not call `parent` directly. Call `$this->validateValue()` and it will call the parent if needed.

Reply:
Plus move this method to the parent `MinimumCharacters` class so you do not need to repeat yourself in every child classes - except `SpecialCharacter` that requires custom handling.